src/Xamarin.Android.Build.Tasks/Utilities/JavaObjectsXmlFile.cs

```diff
@@ -0,0 +1,141 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml;
+using System.Xml.Linq;
+using Java.Interop.Tools.JavaCallableWrappers.Adapters;
+using Java.Interop.Tools.JavaCallableWrappers.CallableWrapperMembers;
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tasks;
+
+class JavaObjectsXmlFile
+{
+	static XmlWriterSettings settings = new XmlWriterSettings {
+		Indent = true,
+		NewLineOnAttributes = false,
+		OmitXmlDeclaration = true,
+	};
+
+	static readonly JavaObjectsXmlFile unscanned = new JavaObjectsXmlFile { WasScanned = false };
+
+	public List<ACWMapEntry> ACWMapEntries { get; } = [];
+	public List<CallableWrapperType> JavaCallableWrappers { get; } = [];
+	public bool WasScanned { get; private set; }
+
+	public void Export (string filename)
+	{
+		using var sw = MemoryStreamPool.Shared.CreateStreamWriter ();
+
+		using (var xml = XmlWriter.Create (sw, settings))
+			Export (xml);
+
+		sw.Flush ();
+
+		Files.CopyIfStreamChanged (sw.BaseStream, filename);
+	}
+
+	void Export (XmlWriter xml)
+	{
+		xml.WriteStartElement ("api");
+
+		ExportJCWTypes (xml);
+		ExportACWMappingTypes (xml);
+
+		xml.WriteEndElement ();
+	}
+
+	void ExportJCWTypes (XmlWriter xml)
+	{
+		if (JavaCallableWrappers.Count == 0)
+			return;
+
+		xml.WriteStartElement ("jcw-types");
+
+		XmlExporter.Export (xml, JavaCallableWrappers);
+
+		xml.WriteEndElement ();
+	}
+
+	void ExportACWMappingTypes (XmlWriter xml)
+	{
+		if (ACWMapEntries.Count == 0)
+			return;
+
+		var t = ACWMapEntries.First ();
+
+		xml.WriteStartElement ("acw-types");
+
+		xml.WriteAttributeStringIfNotDefault ("partial-assembly-name", t.PartialAssemblyName);
+		xml.WriteAttributeStringIfNotDefault ("module-name", t.ModuleName);
+
+		foreach (var type in ACWMapEntries) {
+			xml.WriteStartElement ("type");
+			xml.WriteAttributeStringIfNotDefault ("assembly-qualified-name", type.AssemblyQualifiedName);
+			xml.WriteAttributeStringIfNotDefault ("compat-jni-name", type.CompatJniName);
+			xml.WriteAttributeStringIfNotDefault ("java-key", type.JavaKey);
+			xml.WriteAttributeStringIfNotDefault ("managed-key", type.ManagedKey);
+			xml.WriteAttributeStringIfNotDefault ("partial-assembly-qualified-name", type.PartialAssemblyQualifiedName);
+			xml.WriteEndElement ();
+		}
+
+		xml.WriteEndElement ();
+	}
+
+	/// <summary>
+	/// Given an assembly path, return the path to the ".jlo.xml" file that should be next to it.
+	/// </summary>
+	public static string GetJavaObjectsXmlFilePath (string assemblyPath)
+		=> Path.ChangeExtension (assemblyPath, ".jlo.xml");
+
+	public static JavaObjectsXmlFile Import (string filename, JavaObjectsXmlFileReadType readType)
```

Comment:
API design question: why have the `WasScanned` property at all?  Why not instead have the return type of `Import()` be  (nullable) `JavaObjectsXmlFile?`, with `null` being the "not scanned" value?

Reply:
`WasScanned` feels a little more self-documenting than `null`.  It seems odd to return `null` for "not scanned" and return an empty `JavaObjectsXmlFile` for "scanned, but no data".

Ultimately the issue is probably moot, as the "scanned" concept only exists to facilitate A/B testing between the "current" way and the "new" way (`$(_AndroidJLOCheckedBuild)`).  Once the migration is complete, the old way will be removed, as will the "scanned" concept.